src/Features/LanguageServer/Protocol/Api/ILanguageServerFactory.cs

```diff
@@ -0,0 +1,18 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using StreamJsonRpc;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Api
+{
+    internal interface ILanguageServerFactory
+    {
+        public InProcLanguageServer Create(
+            JsonRpc jsonRpc,
+            ServerCapabilities serverCapabilities,
```

Comment:
So what does it mean to tell create a server and tell it it's capabilities? I would have presumed this was a property that was read back once you created one to know what this server was capable of.

Reply:
While there is an overall set of capabilities the server supports, this allows the caller to specify exactly which ones.  So for example we can incrementally add more features to the extension by allowing the host to add capabilities one at a time

Reply:
So how does the host get the set of capabilities the underlying implementation supports? I guess I'd assume the API is something along the lines of:

1. Ask our library what it supports, you get a ServerCapabitilities back.
2. Turn off what you want to turn off.
3. Pass it back.

Reply:
That's a good question - I think I definitely do need a way to get the set of default capabilities because they would need to get the right set of trigger characters from the completion providers for example.  I may need to expose the capabilities provider.  I'll make that change.

Reply:
Added a capabilities provider and created a default exported value to allow host to get the default capabilities.